074-k8s-storage/proposal.md

```diff
@@ -0,0 +1,167 @@
+# Terms
+- **cache object** a BLOB and relevant metadata that Concourse needs to persist. These could be Resource Caches, Task Caches or Build Caches.
+- **worker** Concourse executes steps on a **worker** and implements some **worker** interface. Concourse is agnostic of the runtime representation of the worker (eg. K8s pod, node or cluster).
+
+# Summary
+
+After spiking on a few solutions for storage on Kubernetes our recommendation is to use an image registry to store **cache objects** for steps.
```

Comment:
it might be interesting to check out the approach that [kaniko](https://github.com/GoogleContainerTools/kaniko) took - it essentially caches layers in a container registry if you don't specify a local path (so, e.g., if you want to run a `Job` that runs `kaniko` but you don't want to care about providing it a volume - which would imply `concurrency=1`)

```console
$ k logs -f build-v1.18.6-4tn7w  | grep cache
INFO[0008] Checking for cached layer cirocosta/envtest-cache:8642ec4096d286832017c667e68548ac74436793bd49b1cf59d6a238044f6b0a...
INFO[0008] No cached layer found for cmd RUN set -x &&          apt update -y && apt install -y rsync &&                git clone https://github.com/kubernetes/kubernetes &&                 cd ./kubernetes &&                      git checkout $KUBE_TAG &&            go mod download &&                       make generated_files
INFO[0194] Checking for cached layer cirocosta/envtest-cache:21931b04c009b9e67db0c6c8a1f907ee699dd32ad3fef2aa14efaf6b1a5dc995...
INFO[0313] Found cached layer, extracting to filesystem
INFO[0348] Checking for cached layer cirocosta/envtest-cache:b862d7fa29205a55ec028679dc2bb416b707a1a29d24feb9714075e4be5a5c74...
INFO[0435] Found cached layer, extracting to filesystem
INFO[0447] Checking for cached layer cirocosta/envtest-cache:e6c3ee315be35a8483933a79682375de049bdef0560721ce9962d728993803d0...
INFO[0464] Found cached layer, extracting to filesystem
INFO[0470] Checking for cached layer cirocosta/envtest-cache:e78396635e5a8c3a4b750b11796109b2a6a954b628c6e24a3dfbc1f579ce7024...
INFO[0471] No cached layer found for cmd RUN go install -v ./cmd/envtest
INFO[0508] Pushing layer cirocosta/envtest-cache:e78396635e5a8c3a4b750b11796109b2a6a954b628c6e24a3dfbc1f579ce7024 to cache now
```

(https://github.com/GoogleContainerTools/kaniko/blob/master/pkg/cache/cache.go)

Reply:
Hey @cirocosta 
We can see this style of local caching being beneficial for pulling the contents on an input for a task. Is that what you meant ? We weren't sure how this might be useful for pushing 🤔

Reply:
oh, I didn't have any specifics in mind, it was more about showcasing the pattern of using a registry as a place where you just throw layers at a registry regardless of what you're building and then try to reuse those layers (pulling if they exist) when running another build